pkg/app/piped/planpreview/builder.go

```diff
@@ -342,23 +343,21 @@ func (b *builder) listApplications(repo config.PipedRepository) []*model.Applica
 }
 
 func (b *builder) getMostRecentlySuccessfulDeployment(ctx context.Context, applicationID string) (*model.ApplicationDeploymentReference, error) {
-	var (
-		err   error
-		resp  *pipedservice.GetApplicationMostRecentDeploymentResponse
-		retry = pipedservice.NewRetry(3)
-		req   = &pipedservice.GetApplicationMostRecentDeploymentRequest{
+	retry := pipedservice.NewRetry(3)
+
+	deploy, err := retry.Do(ctx, func() (interface{}, error) {
+		resp, err := b.apiClient.GetApplicationMostRecentDeployment(ctx, &pipedservice.GetApplicationMostRecentDeploymentRequest{
 			ApplicationId: applicationID,
 			Status:        model.DeploymentStatus_DEPLOYMENT_SUCCESS,
-		}
-	)
-
-	for retry.WaitNext(ctx) {
-		if resp, err = b.apiClient.GetApplicationMostRecentDeployment(ctx, req); err == nil {
+		})
+		if err == nil {
 			return resp.Deployment, nil
 		}
-		if !pipedservice.Retriable(err) {
-			return nil, err
-		}
+		return nil, backoff.NewError(err, pipedservice.Retriable(err))
+	})
+	if err != nil {
+		return nil, err
 	}
-	return nil, err
```

Comment:
The original `WaitNext` function will not execute any RPC if the given context was already canceled then both returned values will be nil and causing a panic.